### PR TITLE
refactor(build): use 'paths' set in TS config in alias plugin

### PR DESCRIPTION
**Pull request by alicewriteswrongs**
This refactors our 'alias' rollup plugin to determine what paths to resolve things to based on the settings we have in `tsconfig.json`.

This is relevant for the mappings that we have set in `paths`, where we map things like `@utils` to `src/utils/index.ts`. TypeScript, because reasons, does not provide an option to automatically resolve those paths to normal relative imports when compiling.

Our solution to this at present is to maintain a list of modules which should be rewritten in this `alias` plugin. However, the information about which module names map to which files is already set in our `tsconfig.json`! So it would be nicer if we didn't have to maintain that information in two places.

This change modifies the `alias` plugin so that it looks up module names in the `paths` object in `tsconfig` and then uses TypeScript's module resolution helper function to ensure it has the right path for the module, and then re-writes the path to point to the right compiled file in the built output.

Note that (unfortunately) this change cannot do the same thing for our `.d.ts` files :( We'll need to figure out another solution to hopefully copy those automatically and re-write import paths within them to point to the right places.

## Pull request checklist

Please check if your PR fulfills the following requirements:
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Build (`npm run build`) was run locally and any changes were pushed
- [x] Unit tests (`npm test`) were run locally and passed
- [x] E2E Tests (`npm run test.karma.prod`) were run locally and passed
- [x] Prettier (`npm run prettier`) was run locally and passed

## Pull request type

 

 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [ ] Feature
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

GitHub Issue Number: N/A

## What is the new behavior?

-
-
-

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -30,6 +30,7 @@
       "@app-globals": ["src/app-globals/index.ts"],
       "@compiler-deps": ["src/compiler/sys/modules/compiler-deps.ts"],
       "@dev-server-process": ["src/dev-server/server-process.ts"],
+      "@environment": ["src/compiler/sys/environment.ts"],
```
Comment (alicewriteswrongs) [inline tsconfig.json:33]: I believe that we programmatically insert an import to this module into TypeScript during the rollup build (so an import from `@environment` doesn't appear anywhere in our code) but if we want it to resolve correctly the easiest / best thing to do is just add it here to our `paths`. We could also then start using this alias throughout our code if we want to.

**Review by alicewriteswrongs — COMMENTED** [review]
just noting one funny bit

**Review by tanner-reits — APPROVED** [review]
This makes sense to me. Some of the existing code for `empty` and `alias` is a bit strange, but if it's been working 🤷‍♂️ These path alias type things keep coming up. I wonder if we'll be able to abstract some of the code between this, the change for `dist-collection` a while back and, #3788

**Comment by alicewriteswrongs** [comment]
yeah, possibly this plugin could be used both to build stencil and by stencil?

**Comment by rwaskiewicz** [comment]
@alicewriteswrongs is this ready to be reviewed/assigned?

**Comment by alicewriteswrongs** [comment]
yes, sorry for not assigning!

**Comment by alicewriteswrongs** [comment]
Just an update on this, I do want to do _something_ along these lines to clear up the `paths`-based module identifiers in the Stencil build, and while I still think this is a better way to do it than what we have in `main` right now I don't think it is \~the way\~, so while I would be fine with merging this for now if the team agrees we could also just stay tuned for further developments for now

**Pull request closed without merging**